README.md
[![npm](https://img.shields.io/npm/v/react-native-banner-lite.svg?style=flat)]()
[![Gemnasium](https://img.shields.io/gemnasium/mathiasbynens/he.svg)]()
[![npm](https://img.shields.io/npm/l/express.svg)]()

# react-native-banner-lite

A paging banner component. Compatible with iOS and Android.

It made from `ScrollView`, but it doesn't depend on the property `paging`. So the banner can be used in all versions of React Native.

## Roadmap

> see: [ROADMAP.md](ROADMAP.md)

## Preview

![](/Preview/p01.gif)

**[Play in the React Native Playground](https://rnplay.org/apps/-Ks6Lg)**

## Getting Started

### Installation

```bash
$ npm i react-native-banner-lite --save
```

### Usage

- Install `react-native` first

```bash
$ npm i react-native -g
```

- Initialization of a react-native project

```bash
$ react-native init myproject
```

- Import `react-native-banner-lite` in the `index.ios.js` or `index.android.js`

```javascript
import BannerLite from 'react-native-banner-lite';
```

- Add the `BannerLite` component to your code

```javascript
<BannerLite
  items={[
    {
      title: "Hello",
      subtitle: "World",
      imageURL: "http://h.hiphotos.baidu.com/image/h%3D200/sign=3a225a4129a4462361caa262a8227246/30adcbef76094b36fbaf3bd6aacc7cd98d109dcf.jpg",
      onPress:(index)=>{console.log("tap"+index)}
    },
    {
      title: "How",
      subtitle: "R U",
      imageURL: "http://a4.att.hudong.com/35/64/01300000276819133197645554930.jpg",
      onPress:(index)=>{console.log("tap"+index)}
    },
    {
      title: "R U OK",
      subtitle: "😂😂😂",
      imageURL: "http://pic69.nipic.com/file/20150610/21067407_235515103000_2.jpg",
      onPress:(index)=>{console.log("tap"+index)}
    },
  ]}
/>
```

- Run the project

```bash
$ react-native run-ios
```

or

```bash
$ react-native run-android
```

See the example code in the Example folder.

### Parameters

| Name  |  Type   | Description        |
| :---- | :-----: | :----------------- |
| items | `array` | banner items array |

The items is an array of objects. The object contains 4 props.

| Name     |   Type   | Description                           |
| :------- | :------: | :------------------------------------ |
| title    | `string` | title of the banner item              |
| subtitle | `string` | subtitle of the banner item           |
| imageURL | `string` | background image of the banner item   |
| onPress  |  `func`  | callback of the banner item tap event |

## Contribution

- [@talisk](mailto:talisk@talisk.cn)

## Questions

Welcome to send Email to me, or open issue on the repository.

The project follow to the MIT license. Welcome contributions.


CHANGELOG.md
- [v1.0.1]
  + Replace detailText by subtitleText.
  + Upgrade examples for iOS and Android.
  + Add the missing semicolon.

- [v1.0.0]
  + Release first version. And open source on [talisk/react-native-banner-lite | GitHub](https://github.com/talisk/react-native-banner-lite)


Example/test.js
import React, {
  Component,
} from 'react';

import {
  PropTypes,
  StyleSheet,
  ScrollView,
  View,
  Dimensions,
  Text,
  Image,
  TouchableOpacity,
} from 'react-native';

const lineHeight = 18;

export default class DiscussBannerScrollView extends Component {

	static propTypes = {
		items: React.PropTypes.arrayOf(React.PropTypes.shape({
      title: React.PropTypes.string,
			subtitle: React.PropTypes.string,
			imageURL: React.PropTypes.string,
			onPress: React.PropTypes.func,
		})),
	};

	constructor(props) {
	  super(props);

	  this.state = {
	  	items: this.props.items,
	  };
	}

	onScrollEndDrag = e => {
    const { contentOffset } = e.nativeEvent;

    if (contentOffset.x < (this.state.items.length-1.8) * (220+12) && (this.state.items.length !== 1 && this.state.items.length !== 0)) {
	    if (contentOffset.x%(220+12) < 116) {
	    	this.refs.scrollView.scrollTo({x: contentOffset.x-contentOffset.x%(220+12), y: 0, animated: true});
	    } else {
	    	this.refs.scrollView.scrollTo({x: (contentOffset.x-contentOffset.x%(220+12))+(220+12), y: 0, animated: true});
	    }
	  } else {
	  	if (contentOffset.x >= (this.state.items.length-1.8) * (220+12) && (this.state.items.length !== 1 && this.state.items.length !== 0)) {
	  		this.refs.scrollView.scrollTo({x: (this.state.items.length*(220+12) - Dimensions.get('window').width)+12, y: 0, animated: true});
	  	} else if (this.state.items.length === 1 || this.state.items.length === 0) {
	  		this.refs.scrollView.scrollTo({x: 0, y: 0, animated: true});
	  	}
	  }
  }

	render() {

		var bannerItems = [];

		for (var i = 0; i < this.state.items.length; i++) {
			bannerItems.push(
				<BannerItem
					key={"bannerItem"+i}
					index={i}
					imageURL={this.state.items[i].imageURL}
          title={this.state.items[i].title}
					subtitle={this.state.items[i].subtitle}
					onPress={this.state.items[i].onPress}
				/>
			);
		}

		if (bannerItems.length === 0) {
			bannerItems.push(
				<View key='empty' style={[itemStyles.containerFirst, itemStyles.itemSize, itemStyles.emptyHolder]}>
					<Text>
						这里什么也没有
					</Text>
				</View>
			);
		}

		if (bannerItems.length === 1) {
			bannerItems.push(
				<View key='holder' style={[itemStyles.container, itemStyles.itemSize, itemStyles.emptyItem]}/>
			);
		}

		return(
			<ScrollView
				ref='scrollView'
				centerContent={false}
				onScrollEndDrag={this.onScrollEndDrag}
				contentContainerStyle={scrollViewStyles.scrollView}
				horizontal={true}
				showsHorizontalScrollIndicator={false}
			>
				{bannerItems}
			</ScrollView>
		);
	}

}

class BannerItem extends Component {

	constructor(props) {
	  super(props);

	  this.state = {
	  	singleLine: false,
	  };
	}

	onTextLayout(event){
		// 计算行高
		let lh = event.nativeEvent.layout.height;
		if (Math.abs(lh - lineHeight) < 2) {
			// 单行
			this.setState({singleLine: true});
		} else {
			// 两行
			this.setState({singleLine: false});
		}
  }

	render() {
		return (
			<TouchableOpacity style={(this.props.index===0)?itemStyles.containerFirst:itemStyles.container} onPress={() => this.props.onPress(this.props.index)}>
				<Image
	        style={[itemStyles.bgImage,itemStyles.itemSize]}
	        source={{uri: this.props.imageURL}}
          borderRadius={8}
	      >
		      <View style={[itemStyles.maskView,itemStyles.itemSize]}>
		      	<Text style={(this.state.singleLine)?[itemStyles.titleText, {marginTop: 26+lineHeight/2}]:itemStyles.titleText} numberOfLines={2} onLayout={this.onTextLayout.bind(this)}>
		      		{this.props.title}
		      	</Text>
		      	<Text style={itemStyles.subtitleText} numberOfLines={1}>
		      		{this.props.subtitle}
		      	</Text>
		      </View>
        </Image>
			</TouchableOpacity>
		);
	}
}

const itemStyles = StyleSheet.create({
	emptyItem: {
		backgroundColor: 'transparent',
	},
	emptyHolder: {
		backgroundColor: 'red',
	},
	container: {
		flex: 1,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		marginRight: 12,
	},
	containerFirst: {
		flex: 1,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		marginRight: 12,
		marginLeft: 12,
	},
	maskView: {
		borderRadius: 8,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'space-between',
		backgroundColor: 'rgba(0,0,0,0.6)',
	},
	bgImage: {
		borderRadius: 8,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
	},
	itemSize: {
		height: 121,
		width: 220,
	},
	titleText: {
		lineHeight: lineHeight,
		marginTop: 26,
		marginLeft: 24,
		marginRight: 24,
		color: 'white',
		backgroundColor: 'transparent',
		fontSize: 14,
		textShadowOffset: {width: 0.1, height: 0.1},
		textShadowColor: "black",
		textShadowRadius: 2.2,
	},
	subtitleText: {
		marginBottom: 26,
		color: 'white',
		backgroundColor: 'transparent',
		fontSize: 12,
		textShadowOffset: {width: 0.1, height: 0.1},
		textShadowColor: "black",
		textShadowRadius: 2.3,
	},
});

const scrollViewStyles = StyleSheet.create({
	scrollView: {
		flexDirection: 'row',
		alignItems: 'center',
	},
});


Example/index.android.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 * @flow
 */

import React, { Component } from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  View
} from 'react-native';

import BannerLite from 'react-native-banner-lite';
// import BannerLite from './test';

export default class Example extends Component {
  render() {
    return (
      <View style={styles.container}>
        <BannerLite
          items={[
            {
              title: "守望先锋的火爆是否意味着MOBA类游戏的衰退期已到",
              subtitle: "呵呵呵呵呵",
              imageURL: "http://imagscdn.3234.com/pictures/0/01/r4UhVf.jpg",
              onPress:(index)=>{console.log("tap"+index)}
            },
            {
              title: "🐶🐶🐶",
              subtitle: "呵呵呵呵呵",
              imageURL: "http://pic69.nipic.com/file/20150610/21067407_235515103000_2.jpg",
              onPress:(index)=>{console.log("tap"+index)}
            },
            {
              title: "独立游戏在中国的发展前景怎么样",
              subtitle: "呵呵呵呵呵",
              imageURL: "http://a4.att.hudong.com/35/64/01300000276819133197645554930.jpg",
              onPress:(index)=>{console.log("tap"+index)}
            },
            {
              title: "魔兽世界",
              subtitle: "呵呵呵呵呵",
              imageURL: "http://pic.58pic.com/58pic/12/56/90/04A58PICsKy.jpg",
              onPress:(index)=>{console.log("tap"+index)}
            },
            {
              title: "哈哈哈哈😜😜😜😜😜😜😜😜😜😜😜😜😜😜😜😜😜",
              subtitle: "呵呵呵呵呵",
              imageURL: "http://h.hiphotos.baidu.com/image/h%3D200/sign=3a225a4129a4462361caa262a8227246/30adcbef76094b36fbaf3bd6aacc7cd98d109dcf.jpg",
              onPress:(index)=>{console.log("tap"+index)}
            },
          ]}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});

AppRegistry.registerComponent('Example', () => Example);


index.js
import React, {
  Component,
} from 'react';

import {
  PropTypes,
  StyleSheet,
  ScrollView,
  View,
  Dimensions,
  Text,
  Image,
  TouchableOpacity,
} from 'react-native';

const lineHeight = 18

export default class DiscussBannerScrollView extends Component {

	static propTypes = {
		items: React.PropTypes.arrayOf(React.PropTypes.shape({
			text: React.PropTypes.string,
			imageURL: React.PropTypes.string,
			joined: React.PropTypes.number,
			countdown: React.PropTypes.number,
			onPress: React.PropTypes.func,
		})),
	};

	constructor(props) {
	  super(props);

	  this.state = {
	  	items: this.props.items,
	  };
	}

	onScrollEndDrag = e => {
    const { contentOffset } = e.nativeEvent;

    if (contentOffset.x < (this.state.items.length-1.8) * (220+12) && (this.state.items.length !== 1 && this.state.items.length !== 0)) {
	    if (contentOffset.x%(220+12) < 116) {
	    	this.refs.scrollView.scrollTo({x: contentOffset.x-contentOffset.x%(220+12), y: 0, animated: true});
	    } else {
	    	this.refs.scrollView.scrollTo({x: (contentOffset.x-contentOffset.x%(220+12))+(220+12), y: 0, animated: true});
	    }
	  } else {
	  	if (contentOffset.x >= (this.state.items.length-1.8) * (220+12) && (this.state.items.length !== 1 && this.state.items.length !== 0)) {
	  		this.refs.scrollView.scrollTo({x: (this.state.items.length*(220+12) - Dimensions.get('window').width)+12, y: 0, animated: true});
	  	} else if (this.state.items.length === 1 || this.state.items.length === 0) {
	  		this.refs.scrollView.scrollTo({x: 0, y: 0, animated: true});
	  	}
	  }
  }

	render() {

		var bannerItems = [];

		for (var i = 0; i < this.state.items.length; i++) {
			bannerItems.push(
				<BannerItem
					key={"bannerItem"+i}
					index={i}
					imageURL={this.state.items[i].imageURL}
					text={this.state.items[i].text}
					joined={this.state.items[i].joined}
					countdown={this.state.items[i].countdown}
					onPress={this.state.items[i].onPress}
				/>
			);
		}

		if (bannerItems.length === 0) {
			bannerItems.push(
				<View key='empty' style={[itemStyles.containerFirst, itemStyles.itemSize, itemStyles.emptyHolder]}>
					<Text>
						这里什么也没有
					</Text>
				</View>
			);
		}

		if (bannerItems.length === 1) {
			bannerItems.push(
				<View key='holder' style={[itemStyles.container, itemStyles.itemSize, itemStyles.emptyItem]}/>
			);
		}

		return(
			<ScrollView
				ref='scrollView'
				centerContent={false}
				onScrollEndDrag={this.onScrollEndDrag}
				contentContainerStyle={scrollViewStyles.scrollView}
				horizontal={true}
				showsHorizontalScrollIndicator={false}
			>
				{bannerItems}
			</ScrollView>
		);
	}

}

class BannerItem extends Component {

	constructor(props) {
	  super(props);

	  this.state = {
	  	singleLine: false,
	  };
	}

	onTextLayout(event){
		// 计算行高
		let lh = event.nativeEvent.layout.height;
		if (Math.abs(lh - lineHeight) < 2) {
			// 单行
			this.setState({singleLine: true});
		} else {
			// 两行
			this.setState({singleLine: false});
		}
  }

	render() {
		return (
			<TouchableOpacity style={(this.props.index===0)?itemStyles.containerFirst:itemStyles.container} onPress={this.props.onPress}>
				<Image
	        style={[itemStyles.bgImage,itemStyles.itemSize]}
	        source={{uri: this.props.imageURL}}
          borderRadius={8}
	      >
		      <View style={[itemStyles.maskView,itemStyles.itemSize]}>
		      	<Text style={(this.state.singleLine)?[itemStyles.titleText, {marginTop: 26+lineHeight/2}]:itemStyles.titleText} numberOfLines={2} onLayout={this.onTextLayout.bind(this)}>
		      		{this.props.text}
		      	</Text>
		      	<Text style={itemStyles.detailText} numberOfLines={1}>
		      		{this.props.joined}人参与·{this.props.countdown}天后结束
		      	</Text>
		      </View>
        </Image>
			</TouchableOpacity>
		);
	}
}

const itemStyles = StyleSheet.create({
	emptyItem: {
		backgroundColor: 'transparent',
	},
	emptyHolder: {
		backgroundColor: 'red',
	},
	container: {
		flex: 1,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		marginRight: 12,
	},
	containerFirst: {
		flex: 1,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		marginRight: 12,
		marginLeft: 12,
	},
	maskView: {
		borderRadius: 8,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'space-between',
		backgroundColor: 'rgba(0,0,0,0.6)',
	},
	bgImage: {
		borderRadius: 8,
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
	},
	itemSize: {
		height: 121,
		width: 220,
	},
	titleText: {
		lineHeight: lineHeight,
		marginTop: 26,
		marginLeft: 24,
		marginRight: 24,
		color: 'white',
		backgroundColor: 'transparent',
		fontSize: 14,
		textShadowOffset: {width: 0.1, height: 0.1},
		textShadowColor: "black",
		textShadowRadius: 2.2,
	},
	detailText: {
		marginBottom: 26,
		color: 'white',
		backgroundColor: 'transparent',
		fontSize: 12,
		textShadowOffset: {width: 0.1, height: 0.1},
		textShadowColor: "black",
		textShadowRadius: 2.3,
	},
});

const scrollViewStyles = StyleSheet.create({
	scrollView: {
		flexDirection: 'row',
		alignItems: 'center',
	},
});
